src/Sarif/Readers/DeferredDictionary.cs

```diff
@@ -0,0 +1,514 @@
+        }
+
+        public void CopyTo(KeyValuePair<string, T>[] array, int arrayIndex)
+        {
+            if (arrayIndex < 0 || arrayIndex + this.Count > array.Length) throw new ArgumentOutOfRangeException("arrayIndex");
+
+            int index = arrayIndex;
+            foreach(KeyValuePair<string, T> item in this)
+            {
+                array[index++] = item;
+            }
+        }
+
+        public IEnumerator<KeyValuePair<string, T>> GetEnumerator()
+        {
+            return new JsonDeferredDictionaryEnumerator<T>(_jsonSerializer, _streamProvider, _start);
+        }
+        
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new JsonDeferredDictionaryEnumerator<T>(_jsonSerializer, _streamProvider, _start);
+        }
+
+        /// <summary>
+        ///  JsonObjectMemberStreamReader is used to get JsonTextReader to read a property name and
+        ///  value somewhere in an outer object.
+        ///  
+        ///  It alters the first read so that there's a fake outer object (a starting '{') and
+        ///  whitespace and any comma before the property name are hidden.
+        /// </summary>
+        private class JsonObjectMemberStreamReader : StreamReader
+        {
+            private bool _readCalled;
+
+            public JsonObjectMemberStreamReader(Stream stream) : base(stream, Encoding.Default, true, 1024, true)
+            { }
+
+            public override int Read(char[] buffer, int index, int count)
+            {
+                int countRead = 0;
+
+                // We are seeking to a position just before a new PropertyName in an object.
+                if(!_readCalled)
+                {
+                    _readCalled = true;
+
+                    // We'll add a '{' to the JsonTextReader understands we're in an object
+                    buffer[index++] = '{';
+
+                    // Read the remaining desired count
+                    countRead = base.Read(buffer, index + 1, count - 1);
+
+                    // Make everything before the next PropertyName whitespace (old space and comma between items)
+                    for(int i = index + 1; i < index + 1 + countRead; ++i)
+                    {
+                        if (buffer[i] == '"') break;
+                        buffer[i] = ' ';
+                    }
+
+                    // Return the length found plus our added '{'
+                    return countRead + 1;
+                }
+
+                return base.Read(buffer, index, count);
+            }
+        }
+
+        /// <summary>
+        ///  JsonDeferredDictionaryEnumerator provides enumeration over a Dictionary in a
+        ///  Json object starting at a given position.
+        ///  
+        ///  Items are not pre-loaded and are not kept after enumeration, allowing use with
+        ///  collections too large for memory.
+        /// </summary>
+        /// <typeparam name="U">Dictionary Value item type</typeparam>
```

Comment:
>U [](start = 29, length = 1)

just curious, why 'U' as a type parameter name?

Reply:
I made this a private class (since it's used for the containing type) and you have to use different type parameter names than the enclosing class or it's unclear to the compiler whether it's a generic class or specific to the particular <T> of the compiled outer class.

(Short answer: You get a compiler warning if you don't.)